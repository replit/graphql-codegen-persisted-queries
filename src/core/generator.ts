import { DocumentNode, visit } from 'graphql';
import { 
  PluginConfig, 
  QueryIdentifierMap 
} from '../types';
import { createHash } from '../utils/hashing';
import { addTypenameToDocument } from '../utils/transforms';
import { findFragments, findUsedFragments } from '../utils/fragments';
import { printDefinitions } from '../utils/transforms';

/**
 * Generates query identifiers for all operations in the provided documents
 * 
 * @param docs - Array of GraphQL document nodes
 * @param config - Plugin configuration
 * @returns Map of operation names to their identifiers
 * @throws Error if an operation is missing a name
 */
export function generateQueryIds(
  docs: DocumentNode[],
  config: PluginConfig,
): QueryIdentifierMap {
  // Add __typename to all selection sets
  const processedDocs = docs.map(addTypenameToDocument);

  const result: QueryIdentifierMap = {};
  const knownFragments = findFragments(processedDocs);

  for (const doc of processedDocs) {
    visit(doc, {
      OperationDefinition: {
        enter(def) {
          if (!def.name) {
            throw new Error('OperationDefinition missing name');
          }

          const operationName = def.name.value;
          const usedFragments = findUsedFragments(def, knownFragments);

          // The order is important here. We want to print the operation definition first,
          // then the fragments as they are included.
          // Otherwise, the generated hashes won't match with the one generated on the server.
          const query = printDefinitions([
            def,
            ...Array.from(usedFragments.values()),
          ]);

          const hash = createHash(query, config);
          const usesVariables = Boolean(
            def.variableDefinitions && def.variableDefinitions.length > 0,
          );

          result[operationName] = {
            hash,
            query,
            usesVariables,
            loc: doc.loc,
          };
        },
      },
    });
  }

  return result;
}