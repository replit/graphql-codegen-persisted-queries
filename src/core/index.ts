export * from './generator';
export * from './manifests';
export * from './plugin';