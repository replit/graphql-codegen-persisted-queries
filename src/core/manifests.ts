import {
  ClientQueryListManifest,
  PersistedQueryManifestOperation,
  QueryIdentifierMap,
  ServerQueryListManifest
} from '../types';

/**
 * Generates a client-side persisted query manifest
 * Client manifests map operation names to their hash
 * 
 * @param queries - Map of queries with their identifiers
 * @returns Client-side manifest object
 */
export function generateClientManifest(
  queries: QueryIdentifierMap,
): ClientQueryListManifest {
  const manifest: ClientQueryListManifest = {};
  
  for (const queryName of Object.keys(queries)) {
    manifest[queryName] = queries[queryName].hash;
  }

  return manifest;
}

/**
 * Generates a server-side persisted query manifest
 * Server manifests map query hashes to operation details
 * 
 * @param queries - Map of queries with their identifiers
 * @returns Server-side manifest object
 * @throws Error if an operation type cannot be determined
 */
export function generateServerManifest(
  queries: QueryIdentifierMap,
): ServerQueryListManifest {
  const manifest: ServerQueryListManifest = {
    format: 'apollo-persisted-query-manifest',
    version: 1,
    operations: {},
  };

  const operations: Record<string, PersistedQueryManifestOperation> = {};
  
  for (const queryName of Object.keys(queries)) {
    const query = queries[queryName].query;
    let type: 'query' | 'mutation' | 'subscription';
    
    // Determine operation type from the query content
    if (query.startsWith('query ')) {
      type = 'query';
    } else if (query.startsWith('mutation ')) {
      type = 'mutation';
    } else if (query.startsWith('subscription ')) {
      type = 'subscription';
    } else {
      throw new Error('Unknown operation type');
    }

    operations[queries[queryName].hash] = {
      type,
      name: queryName,
      body: query,
    };
  }

  manifest.operations = operations;
  return manifest;
}