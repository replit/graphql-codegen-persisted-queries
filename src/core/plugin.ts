import { DocumentNode } from 'graphql';
import { PersistedQueryPlugin } from '../types';
import { generateQueryIds } from './generator';
import { generateClientManifest, generateServerManifest } from './manifests';

/**
 * GraphQL CodeGen plugin for generating persisted query manifests
 * 
 * @param _schema - GraphQL schema (unused)
 * @param documents - GraphQL documents to process
 * @param config - Plugin configuration
 * @returns JSON string representation of the generated manifest
 * @throws Error if no documents are provided or output format is not specified
 * 
 * For GraphQL over HTTP specification compliance, set `includeAlgorithmPrefix: true`
 * to enable the "Prefixed Document Identifier" format (e.g., `sha256:abc123...`).
 */
export const plugin: PersistedQueryPlugin = (
  _schema,
  documents,
  config,
) => {
  // Validate input documents
  if (
    !documents ||
    documents.length === 0 ||
    documents.some((doc) => !doc?.document)
  ) {
    throw new Error('Found no documents to generate persisted query ids.');
  }

  // Extract document nodes
  const documentNodes: DocumentNode[] = documents
    .map((doc) => doc.document)
    .filter((doc): doc is DocumentNode => doc !== undefined);

  // Generate query identifiers
  const queries = generateQueryIds(documentNodes, config);

  // Generate appropriate manifest format based on configuration
  if (config.output === 'client') {
    return JSON.stringify(generateClientManifest(queries), null, '   ');
  } else if (config.output === 'server') {
    return JSON.stringify(generateServerManifest(queries), null, '   ');
  } else {
    throw new Error("Must configure output to 'server' or 'client'");
  }
};