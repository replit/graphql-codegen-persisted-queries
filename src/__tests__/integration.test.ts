import { describe, test, expect } from 'vitest';
import { plugin } from '../core/plugin';
import { parse, buildSchema } from 'graphql';
import { Types } from '@graphql-codegen/plugin-helpers';
import { readFileSync } from 'fs';
import { join } from 'path';

describe('Persisted Query Plugin Integration Tests', () => {
  // Load test schema and documents
  const schemaPath = join(__dirname, 'fixtures/test-schema.graphql');
  const documentsPath = join(__dirname, 'fixtures/test-documents.graphql');

  const schemaDoc = readFileSync(schemaPath, 'utf8');
  const documentsDoc = readFileSync(documentsPath, 'utf8');

  const schema = buildSchema(schemaDoc);
  const documentNode = parse(documentsDoc);

  const documents: Types.DocumentFile[] = [
    {
      document: documentNode,
      location: documentsPath,
    },
  ];

  describe('Client Manifest Generation', () => {
    test('generates valid client manifest with default options', () => {
      const result = plugin(schema, documents, { output: 'client' });
      // We know result is a string, but TypeScript sees it as Promisable<string>
      const manifest = JSON.parse(result as string);

      // Use snapshot for client manifest
      expect(manifest).toMatchSnapshot();
    });

    test('generates valid client manifest with algorithm prefix', () => {
      const result = plugin(schema, documents, {
        output: 'client',
        includeAlgorithmPrefix: true,
      });
      // We know result is a string, but TypeScript sees it as Promisable<string>
      const manifest = JSON.parse(result as string);

      // Use snapshot for client manifest with algorithm prefix
      expect(manifest).toMatchSnapshot();
    });

    test('generates valid client manifest with custom algorithm', () => {
      const result = plugin(schema, documents, {
        output: 'client',
        algorithm: 'md5',
      });
      // We know result is a string, but TypeScript sees it as Promisable<string>
      const manifest = JSON.parse(result as string);

      // Use snapshot for client manifest with custom algorithm
      expect(manifest).toMatchSnapshot();
    });
  });

  describe('Server Manifest Generation', () => {
    test('generates valid server manifest with default options', () => {
      const result = plugin(schema, documents, { output: 'server' });
      // We know result is a string, but TypeScript sees it as Promisable<string>
      const manifest = JSON.parse(result as string);

      // Use snapshot for server manifest
      expect(manifest).toMatchSnapshot();
    });
  });

  describe('Error Handling', () => {
    test('throws error on missing operation names', () => {
      const docWithUnnamedOperation = parse(`
        query {
          hello
        }
      `);

      const docs: Types.DocumentFile[] = [
        {
          document: docWithUnnamedOperation,
          location: 'test.graphql',
        },
      ];

      expect(() => plugin(schema, docs, { output: 'client' })).toThrow(
        'OperationDefinition missing name',
      );
    });
  });
});
