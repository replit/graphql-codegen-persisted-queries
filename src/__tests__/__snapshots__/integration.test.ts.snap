// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Persisted Query Plugin Integration Tests > Client Manifest Generation > generates valid client manifest with algorithm prefix 1`] = `
{
  "CreateUser": "sha256:ba9e16c83890d5d9ecabcb415610364fee2932887b2bc393c7143b71aab269b3",
  "GetPost": "sha256:3c6302a47f6243d487fac04c0af0f5f95e0187c17cb1a27e06c7a7c9f0488fd4",
  "GetPosts": "sha256:ae215a6ca2480a8c5c721801d9f943cbb54055ce3bdf855ef21bd255b6bbfdde",
  "GetUser": "sha256:3ac4d5cb7197d3811eedb53feb1db0615ce67aaa91fe1f8d9175e12d37eb6239",
  "GetUsers": "sha256:85306f39b57686912f3e429274c2448e2b4a94f39293875f42d674a7363574ca",
  "HelloQuery": "sha256:789ccee2f24e6bec795f6d345529a1fc195e282453956a69eef4209b06b347a9",
  "OnUserCreated": "sha256:3af4547b737d2086c4a7f6f74f61e310c1c9d1a54c6218f34edfa592a43bf7b5",
  "UpdateUser": "sha256:2ed387ebd9ffea9ea9b42fb1c4c2d149d3c5c45c2c8603ba4ffc755d296dc794",
}
`;

exports[`Persisted Query Plugin Integration Tests > Client Manifest Generation > generates valid client manifest with custom algorithm 1`] = `
{
  "CreateUser": "08382963981d862dd2fd2bfe0e9cd207",
  "GetPost": "79c0b9239ad347d5710bf729c8fdfe66",
  "GetPosts": "69714bf6547e845a36de493b21f754c8",
  "GetUser": "0b03e148029a42a67ed0f9c84b96c1ad",
  "GetUsers": "b584a3ef5982a75fa6ea872c867c4a53",
  "HelloQuery": "7a65e58a5f2108e855f55246308323a3",
  "OnUserCreated": "5cdc0f7772d519dd6543d8c1858f0a79",
  "UpdateUser": "fdec2e64f7bea659f50032796d8e84a6",
}
`;

exports[`Persisted Query Plugin Integration Tests > Client Manifest Generation > generates valid client manifest with default options 1`] = `
{
  "CreateUser": "ba9e16c83890d5d9ecabcb415610364fee2932887b2bc393c7143b71aab269b3",
  "GetPost": "3c6302a47f6243d487fac04c0af0f5f95e0187c17cb1a27e06c7a7c9f0488fd4",
  "GetPosts": "ae215a6ca2480a8c5c721801d9f943cbb54055ce3bdf855ef21bd255b6bbfdde",
  "GetUser": "3ac4d5cb7197d3811eedb53feb1db0615ce67aaa91fe1f8d9175e12d37eb6239",
  "GetUsers": "85306f39b57686912f3e429274c2448e2b4a94f39293875f42d674a7363574ca",
  "HelloQuery": "789ccee2f24e6bec795f6d345529a1fc195e282453956a69eef4209b06b347a9",
  "OnUserCreated": "3af4547b737d2086c4a7f6f74f61e310c1c9d1a54c6218f34edfa592a43bf7b5",
  "UpdateUser": "2ed387ebd9ffea9ea9b42fb1c4c2d149d3c5c45c2c8603ba4ffc755d296dc794",
}
`;

exports[`Persisted Query Plugin Integration Tests > Server Manifest Generation > generates valid server manifest with default options 1`] = `
{
  "format": "apollo-persisted-query-manifest",
  "operations": {
    "2ed387ebd9ffea9ea9b42fb1c4c2d149d3c5c45c2c8603ba4ffc755d296dc794": {
      "body": "mutation UpdateUser($id: ID!, $name: String, $email: String) {
  updateUser(id: $id, name: $name, email: $email) {
    ...UserFields
    __typename
  }
}

fragment UserFields on User {
  id
  name
  email
  __typename
}",
      "name": "UpdateUser",
      "type": "mutation",
    },
    "3ac4d5cb7197d3811eedb53feb1db0615ce67aaa91fe1f8d9175e12d37eb6239": {
      "body": "query GetUser($id: ID!) {
  user(id: $id) {
    ...UserFields
    posts {
      ...PostFields
      __typename
    }
    __typename
  }
}

fragment UserFields on User {
  id
  name
  email
  __typename
}

fragment PostFields on Post {
  id
  title
  content
  __typename
}",
      "name": "GetUser",
      "type": "query",
    },
    "3af4547b737d2086c4a7f6f74f61e310c1c9d1a54c6218f34edfa592a43bf7b5": {
      "body": "subscription OnUserCreated {
  userCreated {
    ...UserFields
    __typename
  }
}

fragment UserFields on User {
  id
  name
  email
  __typename
}",
      "name": "OnUserCreated",
      "type": "subscription",
    },
    "3c6302a47f6243d487fac04c0af0f5f95e0187c17cb1a27e06c7a7c9f0488fd4": {
      "body": "query GetPost($id: ID!) {
  post(id: $id) {
    ...PostFields
    author {
      ...UserFields
      __typename
    }
    __typename
  }
}

fragment PostFields on Post {
  id
  title
  content
  __typename
}

fragment UserFields on User {
  id
  name
  email
  __typename
}",
      "name": "GetPost",
      "type": "query",
    },
    "789ccee2f24e6bec795f6d345529a1fc195e282453956a69eef4209b06b347a9": {
      "body": "query HelloQuery {
  hello
}",
      "name": "HelloQuery",
      "type": "query",
    },
    "85306f39b57686912f3e429274c2448e2b4a94f39293875f42d674a7363574ca": {
      "body": "query GetUsers {
  users {
    ...UserFields
    __typename
  }
}

fragment UserFields on User {
  id
  name
  email
  __typename
}",
      "name": "GetUsers",
      "type": "query",
    },
    "ae215a6ca2480a8c5c721801d9f943cbb54055ce3bdf855ef21bd255b6bbfdde": {
      "body": "query GetPosts {
  posts {
    ...PostFields
    __typename
  }
}

fragment PostFields on Post {
  id
  title
  content
  __typename
}",
      "name": "GetPosts",
      "type": "query",
    },
    "ba9e16c83890d5d9ecabcb415610364fee2932887b2bc393c7143b71aab269b3": {
      "body": "mutation CreateUser($name: String!, $email: String) {
  createUser(name: $name, email: $email) {
    ...UserFields
    __typename
  }
}

fragment UserFields on User {
  id
  name
  email
  __typename
}",
      "name": "CreateUser",
      "type": "mutation",
    },
  },
  "version": 1,
}
`;
